### DISCUSSION TITLE
Printing in uno

### DISCUSSION THREAD

**Discussion by DavidM29** [unoplatform/uno#18623] [Q&A] [answered]
Is it possible to convert base64string to a pdf file and print it on a local network printer from uno?
I know how how to create and save a pdf file from base64string, however I could not find a way to print the said pdf file, [this example that I found](https://learn.microsoft.com/en-us/windows/apps/develop/devices-sensors/print-from-your-app) seems to only print xaml or html pages and I was not able to find anything other that that.

Also this might not be the brightest question but do I have to save the file on the device or can I just create it in the memory and print it?

**Comment by DavidM29** [comment, marked as answer]
So, I think I asked the question too early without doing enough research, however if anyone needs to print pdf documents from winui, these are the articles that you need:
[Printing from winui](https://learn.microsoft.com/en-us/windows/apps/develop/devices-sensors/print-from-your-app)
[Display pdf in xaml](https://stackoverflow.com/questions/78929512/how-display-pdf-file-in-winui-3-desktop-app)

To briefly summarize, you need to create a pdf document from a stream, then create a bitmap image from every single page in the pdf and after that you can create Image controls from these bitmap images and print them according to the Microsoft article I mentioned above.

**Comment by SongOfYouth** [comment]
Have you ever print it on wasm? @DavidM29

**Reply by DavidM29** [reply]
No, I do not target wasm, however, as I checked, a lot of printing related functionality is not implemented in uno so I do not think you can use this on wasm.
